phpBB/phpbb/storage/storage.php

```diff
@@ -133,11 +140,64 @@ public function rename($path_orig, $path_dest)
 	 * @param string	$path_orig	The original filename
 	 * @param string	$path_dest	The target filename
 	 *
-	 * @throws \phpbb\storage\exception\exception	When target exists
-	 * 												When the file cannot be copied
+	 * @throws \phpbb\storage\exception\exception		When target exists
+	 * 													When the file cannot be copied
+	 * @throws \phpbb\storage\exception\not_implemented	When the adapter doesnt implement the method
 	 */
 	public function copy($path_orig, $path_dest)
 	{
 		$this->get_adapter()->copy($path_orig, $path_dest);
 	}
+
+	/**
+	 * Reads a file as a stream.
+	 *
+	 * @param string	$path	File to read
+	 *
+	 * @throws \phpbb\storage\exception\exception		When cannot open file
+
+	 * @return resource	Returns a file pointer
+	 */
+	public function read_stream($path)
+	{
+		$stream = null;
+		$adapter = $this->get_adapter();
+
+		if ($adapter instanceof stream_interface)
+		{
+			$stream = $adapter->read_stream($path);
+		}
+		else
+		{
+			// Simulate the stream
+			$stream = fopen('php://temp', 'w+b');
+			fwrite($stream, $adapter->get_contents($path));
+			rewind($stream);
+		}
+
+		return $stream;
+	}
+
+	/**
+	 * Writes a new file using a stream.
+	 *
+	 * @param string	$path		The target file
+	 * @param resource	$resource	The resource
+	 *										When target file cannot be created
+	 * @throws \phpbb\storage\exception\not_implemented	When the adapter doesnt implement the method
```

Comment:
but gravatar can't be seen as a storage backend. For me storage should all be read/write. Maybe later we could support read only backend with a few changes (but I think it's not important for now and it should be done specifically otherwise it can be confusing)